W-14927671  Pdk ga branch gr (#464)
mulesoft/docs-gateway: pdk/1.0/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc

= Defining a Policy Schema Definition
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

Similar to the policies listed in the xref:gateway::policies-included-directory.adoc[], custom policies have configuration parameters. Use the configuration parameters to improve the usability of your custom policy and support a wider range of API configurations.

> To define custom configuration parameters, edit the `definition/gcl.yaml` file. The `gcl.yaml` file is written in Gateway Configuration Language (GCL). However, you do not need prior GCL knowledge to use this tutorial.

Reviewer: ```suggestion
To define custom configuration parameters, edit the `definition/gcl.yaml` file. The `gcl.yaml` file is written in the Gateway Configuration Language (GCL). However, you do not need prior GCL knowledge to use this tutorial.
```
Not sure how marketing wants this positioned (with or without the "the"?)

The configuration parameters you define in the `gcl.yaml` appear as variables in the `src/generated/config.rs` file after you run the `build-asset-file` command. You set any user input you need for implementing the Rust source code in the `gcl.yaml` file.

Configuration parameters appear in the API Manager UI when applying a policy to Flex Gateway running in Connected Mode. When applying the policy to Flex Gateway running in Local Mode, it is your responsibility to distribute a configuration `yaml` file for your policy.

> The default `gcl.yaml` file has the following content:

Reviewer: You mention above that this is a tutorial, but there aren't any numbered steps to follow. Are these all optional or required?

[source,yaml]
----
apiVersion: gateway.mulesoft.com.v1alpha1
kind: Extension
metadata:
    labels:
        title: my-custom-policy
        category: Custom
spec:
    extends:
        - name: extension-definition
    properties:
        stringProperty:
        type: string
----

The `gcl.yaml` file contains two sections:

* `metadata`: Defines information about the policy
* `properties`: Defines the policy's configuration parameters

To define a schema definition:

. <<define-metadata>>
. <<define-parameters>>
. <<use-custom-parameters-in-rust-source-code>>

[[define-metadata]]
== Define Metadata

Add labels in the metadata section to enrich the data shown in the UI when applying the policy.

> To define the metadata, edit the `metadata` in the `gcl.yaml` file.

Reviewer: Numbered step?

The supported labels are:

* `title`: The policy name that appears in the UI.
* `description`: A brief explanation of the policy's behavior.
* `category`: The category the policy appears under. Supported values are: `security`, `compliance`, `transformation`, `quality of service`, and `troubleshooting`.

For example:

[source,yaml]
----
metadata:
    labels:

>         title: Json Threat Protection

Reviewer: ```suggestion
        title: JSON Threat Protection
```

        description: Protects against malicious JSON in API requests.
        category: Security
----

`labels` are suggested but not required.

[[define-parameters]]
== Define Parameters

> To add parameters, edit the `spec.properties` section in the `gcl.yaml` file.

Reviewer: Numbered step?

For example:

[source,yaml]
----
properties:
    username:
        type: string
    password:
        type: string
----

> After defining your properties, you can:

Reviewer: Numbered step? Is it required or optional?

* <<define-required-properties-and-default-values>>
* <<define-supported-types>>

=== Define Required Properties and Default Values

You can define each property listed in the `gcl.yaml` file as `required` and set a `default` value for the property.

To define properties as required, list them in the `spec.required` object. For example, the following policy requires a `password` but not a `username`:

[source,yaml]
----
properties:
    username:
        type: string
    password:
        type: string
required:
    - password
----

> In this configuration example, applying a policy without a password results in a failed configuration.

Reviewer: Is "this" referring to the above or below configuration example? Or both?

Author: https://github.com/mulesoft/docs-gateway/pull/504

The following example defines a default value for each property:

[source,yaml]
----
properties:
    username:
        type: string
        default: user1
    password:
        type: string
        default: 12345678
required:
    - password
----

> If you don't set a value for a required property when applying a policy, the property uses the default value.

Reviewer: ```suggestion
If you do not set a value for a required property when applying a policy, the property uses the default value.
```

> For non-required properties, the API Manager UI suggests the configured default value for properties. If you don't specify the value for non-required properties, the property is not configured.

Reviewer: ```suggestion
For non-required properties, the API Manager UI suggests the configured default value for properties. If you do not specify the value for non-required properties, the property is not configured.
```

> In the following configuration table, if the policy was applied in API Manager, `user2` would appear as a suggested value for `username` but does not affect if the value is configured:

Reviewer: ```suggestion
In the following configuration table, if the policy was applied in API Manager, `user2` would appear as a suggested value for `username` but is not affected if the value is configured:
```
I'm not sure what the intended meaning is here? What is not affected?

Author: https://github.com/mulesoft/docs-gateway/pull/504

[a,a,a]
|===
| Policy Configuration | Effective Configuration | Configuration Status
| `{}` | `{"password": "12345678"}` | Success
| `{"username": "user2"}` | `{"password": "12345678", +
"username": "user2"}` | Success
| `{"password": "password"}` | `{"password": "password"}` | Success
| `{"password": "password", +
 "username": "user2"}` | `{"password": "password", +
"username": "user2"}` | Success
|===

=== Define Sensitive Data

Protect sensitive data, such as passwords or certificates, by adding the `security:sensitive` characteristic:

[source,yaml]
----
properties:
    key:
        type: string
        default: "YourKey"
        "@context": {
        "@characteristics": [
          "security:sensitive"
        ]
      }
----

When applying the policy, the input value to the sensitive parameter is not visible to the user:

image:sensitive-data.png[The character inputs to the key parameter appear as black dots.]

> Toggling *Show* reveals any new data currently being entered. However, you cannot reveal previously stored data. For example, if a user configures the example *Key* parameter and applies the policy, they cannot view the parameter if they edit the policy's configuration in the future.

Reviewer: ```suggestion
Toggling *Show* reveals any new data currently being entered. However, you cannot reveal previously stored data. For example, if you configure the example *Key* parameter and apply the policy, you cannot view the parameter if you edit the policy's configuration in the future.
```

=== Define Supported Types

Define a supported data type for each policy parameter to ensure the policy receives the correct data.

The supported types are:

String::
+
The string type specifies that the data is text.
+
[source,yaml]
----
properties:
    password:
        type: string
----
+
The `string` type also supports enables you to configure a `format` type.
+
The supported formating types are:
+
* `dataweave`: Specifies the input parameter is a DataWeave variable and notifies Flex Gateway to transform the input. For more information about DataWeave expressions, see <<dataweave-expressions>>.

> * `service`: Specifies the input parameter is an URI and automatically creates a Flex Gateway service to enable HTTP calls from the policy.

Reviewer: ```suggestion
* `service`: Specifies the input parameter is a URI and automatically creates a Flex Gateway service to enable HTTP calls from the policy.
```

> * `ip-range`: Specifies the expected text is a range or IPs and provides additional validation on the input.

Reviewer: ```suggestion
* `ip-range`: Specifies the expected text is a range of IPs and provides additional validation on the input.
```

> * `uri`: Specifies the expected text is an URI and provides additional validation on the input. Only use `uri` when the uri is not a service making HTTP calls from the policy.

Reviewer: ```suggestion
* `uri`: Specifies the expected text is a URI and provides additional validation on the input. Only use `uri` when the URI is not a service making HTTP calls from the policy.
```

=== DataWeave Expressions

The `dataweave` format for `string` type parameters specifies the input parameter as a DataWeave expression.

The `dataweave` variable supports the following expression bindings:

[source,yaml]
----
<configuration-parameter>:
 type: string
 format: dataweave
 bindings:
   payload:
     mimeType: text
   attributes: true
   authentication: false
   vars: []
----

* `payload`: The body of the incoming request. Define the `mimeType` as either `text` or `json` to set how to read the payload. By default, `mimeType` is set to `text`. To restrict the use of the `payload` variable, you must exclude the payload variable, for example:
+
[source,yaml]
----
bindings:
   attributes: true
   authentication: false
   vars: []
----

* `attributes`: Attributes, such as headers, of the incoming request. Set `attributes` to `true` to enable the policy to read the attributes of the incoming request. By default, `attributes` is set to `true`.
* `authentication`: Authentication information of the incoming request. Set `authentication` to `true` to enable the policy to read authentication details of the incoming request. By default, `authentication` is set to `false`.

> * `vars`: Specifies the names of the variables that are available for the policy to use. By default, `vars` is empty. If the user inputs a variable that is not defined in the schema, the policy evaluates the variable as `null`.

Reviewer: Who is "the user" here? The developer?

Author: https://github.com/mulesoft/docs-gateway/pull/504

For more information about the supported expression bindings, see xref:dataweave::dataweave-variables-context.adoc[].

> If the user's expression contains bindings that are not defined in the policy's schema, policy configuration fails. For example, if the user expression is `[authentication.principal]` and `authentication` is not set to `true`, an error similar to the following occurs:

Reviewer: Who is the user here? The developer?

Author: https://github.com/mulesoft/docs-gateway/pull/504

----
test-local-flex-1  | [flex-gateway-envoy][error] wasm log main: [policy: ingress-http-jwt-auth-v1-0-impl-1.default][api: ingress-http.default.svc] Launcher problem: Failed to parse configuration '{"dw":"P[[\".\", \"0-24\", [\":ref\", \"0-14\", \"authentication\"], [\":str\", \"15-24\", \"principal\"]], \"#[authentication.principal]\"]"}'. Cause: Error compiling expression: Unknown symbol `authentication` at line 1 column 141
----

When evaluating the DataWeave expression in your policy, you must provide the script to resolve the expression. To learn how to implement DataWeave expressions in your Rust source code, see xref:policies-pdk-configure-features.adoc#use-dataweave-expressions[Implement DataWeave Expressions].

For more information about DataWeave support for Flex Gateway, see xref:gateway::policies-flex-dataweave-support.adoc[].

[[use-custom-parameters-in-rust-source-code]]
== Use Custom Parameters in Rust Source Code

> After you modify the `gcl.yaml` file, run the `build-asset-files` command to propagate the changes to the `src/generated/config.rs` rust file and the JSON schema for the UI:

Reviewer: Numbered step?

Author: Ask if there’s helps
